File: 316. Remove Duplicate Letters/C++/Remove Duplicate Letters.cpp
class Solution {
public:
    string removeDuplicateLetters(string s) {
        stack<char> sta;
        vector<int> counts(128, 0);
        for (char letter : s) {
            counts[letter]++;
        }
        vector<bool> exists(128, false);
        for (auto letter : s) {
            counts[letter]--;
            if (exists[letter]) {
                continue;
            }
            while (!sta.empty() && letter < sta.top() && counts[sta.top()] > 0) {
                int temp = sta.top();
                sta.pop();
                exists[temp] = false;
            }
            sta.push(letter);
            exists[letter] = true;
        }
        string result;
        while (!sta.empty()) {
            result = sta.top() + result;
            sta.pop();
        }
        return result;
    }
};

File: 316. Remove Duplicate Letters/Java/Remove Duplicate Letters.java
public class Solution {
    public String removeDuplicateLetters(String s) {
        Stack<Character> stack = new Stack<>();
        int[] counts = new int[128];
        for (char letter : s.toCharArray()) {
            counts[letter]++;
        }
        boolean[] exists = new boolean[128];
        for (char letter : s.toCharArray()) {
            counts[letter]--;
            if (exists[letter]) {
                continue;
            }
            while (!stack.empty() && letter < stack.peek() && counts[stack.peek()] > 0) {
                int temp = stack.pop();
                exists[temp] = false;
            }
            stack.push(letter);
            exists[letter] = true;
        }
        StringBuilder builder = new StringBuilder();
        for (char letter : stack) {
            builder.append(letter);
        }
        return builder.toString();
    }
}